core/src/heaviest_subtree_fork_choice.rs

```diff
@@ -744,29 +747,14 @@ impl ForkChoice for HeaviestSubtreeForkChoice {
         &mut self,
         bank: &Bank,
         _tower: &Tower,
-        progress: &mut ProgressMap,
-        computed_bank_state: &ComputedBankState,
+        latest_validator_votes_for_frozen_banks: &mut LatestValidatorVotesForFrozenBanks,
     ) {
-        let ComputedBankState { pubkey_votes, .. } = computed_bank_state;
         let mut start = Measure::start("compute_bank_stats_time");
         // Update `heaviest_subtree_fork_choice` to find the best fork to build on
         let root = self.root.0;
+        let new_votes = latest_validator_votes_for_frozen_banks.take_votes_dirty_set(root);
```

Comment:
As discussed, I think the current behavior in that case is fine.

The opposite ordering of events though, can be kind of tricky. What if a root bank is set, purges a bigger frozen, but unrooted slot, and then a vote comes in for that slot? The nodes that purged will process that vote, the others won't

Reply:
What if I added a new rule that once  the branch you're considering adds a slot greater than the unrooted vote, then we remove the smaller vote? For instance

```
                           |----------------------5 (36% voted here). ------- 8 (Added 8 > 6 which is the newer vote, so we can remove the 2% vote on 5)
  -------  1  ----2 ---3 (new root)
 |                        |------------------------7 (37% voted here)
 |
0
  |------6 (2% new vote)
```

Reply:
Way I think this would be implemented. When we do aggregation, we also propagate up the smallest leaf.  When we calculate the weight of a node `N`, if there is a  validator `V` who's latest frozen vote is for `N`, then we count validator `V`'s stake toward `N` iff the smallest leaf in the subtree of `N` is less than the smallest *unfrozen/unrooted* vote `U_v` made by `V` where `U_v > N`

Reply:
I don't think we can reason about unfrozen slot hashes at all here. Every after adding 8, if 6 has a vote and is unfrozen, it could still chain back to 5.

I think we can only consider frozen slot hashes, if we know a vote is for a frozen slot and that slot is no longer in fork choice and it's for a larger slot than the latest vote, we can remove the stake for the latest vote.

Reply:
> I don't think we can reason about unfrozen slot hashes at all here. Every after adding 8, if 6 has a vote and is unfrozen, it could still chain back to 5.
> 
> I think we can only consider frozen slot hashes, if we know a vote is for a frozen slot and that slot is no longer in fork choice and it's for a larger slot than the latest vote, we can remove the stake for the latest vote.

Yeah so if 6 chains to 5, then we when we add 6 to the fork choice tree, we will also add the newest latest votes from the "dirt set", so it will count that stake toward the ancestors.